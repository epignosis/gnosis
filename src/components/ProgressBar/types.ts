export type Size = "md" | "lg" | "sm" | "xs" | number;
export type Color = "primary" | "success" | "white" | "darkgreen";
export type BorderRadius = {
  startStartRadius: number;
  startEndRadius: number;
  endEndRadius: number;
  endStartRadius: number;
};
