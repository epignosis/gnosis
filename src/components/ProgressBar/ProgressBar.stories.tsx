import React, { ReactNode } from "react";
import { Story } from "@storybook/react";
import ProgressBar, { ProgressBarProps } from "./ProgressBar";

export default {
  component: ProgressBar,
  title: "Components/ProgressBar",
  argTypes: {
    percent: {
      control: {
        type: "range",
        min: 0,
        max: 100,
        step: 1,
      },
    },
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
  },
  decorators: [
    (story: () => ReactNode): JSX.Element => <div style={{ maxWidth: 500 }}>{story()}</div>,
  ],
};

export const Default: Story<ProgressBarProps> = (args) => <ProgressBar {...args} />;

Default.args = {
  size: "md",
  percent: 75,
};
