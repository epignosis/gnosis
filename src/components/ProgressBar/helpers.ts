import { BorderRadius } from "./types";

export function getBorderRadiusCss(borderRadius?: number | BorderRadius): string {
  if (typeof borderRadius === "number") {
    return `border-radius: ${borderRadius}px;`;
  }

  if (borderRadius && typeof borderRadius === "object") {
    return `
   border-start-start-radius: ${borderRadius.startStartRadius}px;
   border-start-end-radius: ${borderRadius.startEndRadius}px;
   border-end-end-radius: ${borderRadius.endEndRadius}px;
   border-end-start-radius: ${borderRadius.endStartRadius}px;`;
  }

  return "";
}
