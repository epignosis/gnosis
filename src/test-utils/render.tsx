import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
