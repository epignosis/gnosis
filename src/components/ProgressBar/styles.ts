import { css, Theme, SerializedStyles } from "@emotion/react";
import { Size } from "./ProgressBar";

const progressBarSize = {
  md: "28px",
  lg: "40px",
};

export const container = (
  { progressBar }: Theme,
  { percent, size, rounded }: { percent: number; size: Size; rounded: boolean },
): SerializedStyles => {
  return css`
    position: relative;
    height: ${progressBarSize[size]};
    width: 100%;
    background-color: ${progressBar.background};
    ${rounded && "border-radius: 5px"};
    overflow: hidden;

    &::after {
      content: " ";
      position: absolute;
      top: 0;
      inset-inline-start: 0;
      height: 100%;
      min-width: ${percent ? "20%" : "0%"};
      width: ${percent}%;
      background-color: ${progressBar.progressBackground};
    }

    div {
      position: absolute;
      z-index: 1;
      height: 100%;
      width: ${percent - 5}%;
      max-width: 95%;
      min-width: ${percent ? "13%" : "0%"};
      display: flex;
      padding-inline-start: 1rem;
      justify-content: flex-end;
      align-items: center;
      color: ${progressBar.percentColor};
    }
  `;
};
