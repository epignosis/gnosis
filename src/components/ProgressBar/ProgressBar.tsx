import React, { FC } from "react";
import { SerializedStyles } from "@emotion/react";
import { container } from "./styles";

export type Size = "md" | "lg";

export type ProgressBarProps = {
  percent: number;
  size?: Size;
  rounded?: boolean;
};

const ProgressBar: FC<ProgressBarProps> = ({ percent, size = "md", rounded = true }) => {
  const showPercentage = percent >= 0 && percent <= 100;

  return (
    <div css={(theme): SerializedStyles => container(theme, { percent, size, rounded })}>
      {showPercentage && <div>{percent}%</div>}
    </div>
  );
};

export default ProgressBar;
