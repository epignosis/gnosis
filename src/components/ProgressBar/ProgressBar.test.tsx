import React from "react";
import { faker } from "@faker-js/faker";
import ProgressBar from "./ProgressBar";
import { render, screen } from "@test-utils/render";

describe("<ProgressBar />", () => {
  it("renders correctly", () => {
    const progressPercentage = faker.datatype.number(100);

    render(<ProgressBar percent={progressPercentage} />);
    const percentage = screen.getByText(`${progressPercentage}%`);

    expect(percentage).toHaveTextContent(`${progressPercentage}%`);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <ProgressBar id="my-progress-1" className="progress" percent={75} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with zero percent", () => {
    const { container } = render(<ProgressBar percent={0} />);

    expect(container).toMatchSnapshot();
  });
});
