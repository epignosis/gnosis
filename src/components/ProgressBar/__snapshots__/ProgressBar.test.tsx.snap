// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> matches snapshot 1`] = `
<div>
  <div
    class="progress css-ocyx8f-container-container"
    id="my-progress-1"
  >
    <div>
      75
      %
    </div>
  </div>
</div>
`;

exports[`<ProgressBar /> matches snapshot with zero percent 1`] = `
<div>
  <div
    class="css-f3opwn-container-container"
  >
    <div>
      0
      %
    </div>
  </div>
</div>
`;
